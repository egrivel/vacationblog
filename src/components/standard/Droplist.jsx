import React from 'react';
import PropTypes from 'prop-types';

import Formrow from './Formrow';

class Droplist extends React.Component {
  static propTypes = {
    fieldId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    list: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })).isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string
  }

  _onChange(event) {
    this.props.onChange(event.target.value, this.props.fieldId);
  }

  render() {
    const options = [];
    const value = this.props.value || '';
    const {list} = this.props;

    for (let i = 0; i < list.length; i++) {
      const item = list[i];
      options.push(
        <option
          key={'o-' + i}
          value={item.value}
        >
          {item.label}
        </option>
      );
    }

    return (
      <Formrow
        key={'k-' + this.props.fieldId}
        label={this.props.label}
        labelFor={this.props.fieldId}
      >
        <select
          id={this.props.fieldId}
          onChange={event => this._onChange(event)}
          value={value}
        >
          {options}
        </select>
      </Formrow>
    );
  }
}

export default Droplist;
