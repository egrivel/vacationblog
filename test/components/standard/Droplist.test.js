import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import enzyme from 'enzyme';

import Droplist from '../../../src/components/standard/Droplist';

/*
 * Test strategy
 * The component has one optional prop (value) and one callback
 * function (onChange). The requirements are:
 *  - the component renders with or without a value
 *  - the onChange is fired when changing the selected item
 */
describe('components/standard/Droplist', () => {
  const testFieldId = 'test-field-id';
  const testLabel = 'Test Label';
  const testItemLabel1 = 'Test Item Label 1';
  const testItemValue1 = 'test-item-value-1';
  const testItemLabel2 = 'Test Item Label 2';
  const testItemValue2 = 'test-item-value-2';
  const testItemLabel3 = 'Test Item Label 3';
  const testItemValue3 = 'test-item-value-3';
  let onChangeStub;
  let testEvent;
  let testList;

  beforeEach(() => {
    onChangeStub = sinon.stub();
    testEvent = {
      preventDefault() {},
      target: {}
    };
    testList = [
      {
        label: testItemLabel1,
        value: testItemValue1
      }, {
        label: testItemLabel2,
        value: testItemValue2
      }, {
        label: testItemLabel3,
        value: testItemValue3
      }
    ];
  });

  afterEach(() => {
    sinon.restore();
  });

  describe('#render', () => {
    test('renders without a value', () => {
      const render = renderer.create(
        <Droplist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });

    test('renders checked', () => {
      const render = renderer.create(
        <Droplist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
          value={testItemValue2}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });
  });

  describe('handles onChange', () => {
    it('passes the selected value back', () => {
      const component = enzyme.shallow(
        <Droplist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
        />
      );
      testEvent.target.value = testItemValue1;
      component.find('select').simulate('change', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      expect(onChangeStub.args[0][0]).toBe(testItemValue1);
      expect(onChangeStub.args[0][1]).toBe(testFieldId);
    });
  });
});
