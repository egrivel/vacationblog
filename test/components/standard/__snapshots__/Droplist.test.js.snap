// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/standard/Droplist #render renders checked 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <select
      id="test-field-id"
      onChange={[Function]}
      value="test-item-value-2"
    >
      <option
        value="test-item-value-1"
      >
        Test Item Label 1
      </option>
      <option
        value="test-item-value-2"
      >
        Test Item Label 2
      </option>
      <option
        value="test-item-value-3"
      >
        Test Item Label 3
      </option>
    </select>
  </div>
</div>
`;

exports[`components/standard/Droplist #render renders without a value 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <select
      id="test-field-id"
      onChange={[Function]}
      value=""
    >
      <option
        value="test-item-value-1"
      >
        Test Item Label 1
      </option>
      <option
        value="test-item-value-2"
      >
        Test Item Label 2
      </option>
      <option
        value="test-item-value-3"
      >
        Test Item Label 3
      </option>
    </select>
  </div>
</div>
`;
